db/db_sst_test.cc

```diff
@@ -1125,6 +1125,9 @@ TEST_F(DBSSTTest, DBWithMaxSpaceAllowedWithBlobFiles) {
 
   TEST_SYNC_POINT("DBSSTTest::DBWithMaxSpaceAllowedWithBlobFiles:1");
   ASSERT_TRUE(delete_blob_file);
+
+  sfm->Close();
```

Comment:
Not entirely sure about this... Looking at the other test cases, there doesn't seem to be a need in general to `Close` the `SstFileManager`.

I do see an issue on line 1100 though: I believe we would really want to call `sfm->SetMaxAllowedSpaceUsage` with `total_files_size + 1` there (not `files_size + 1`). Maybe `ClearError` is getting stuck in a loop because of that?

Reply:
BTW, have you been able to reproduce the race locally? It would be nice to confirm the issue and the effectiveness of any fix. The way to do that would be:

```
make clean
COMPILE_WITH_TSAN=1 make db_sst_test -j24
./db_sst_test --gtest_filter="* DBWithMaxSpaceAllowedWithBlobFiles" --gtest_repeat=1000
```

If it doesn't repro, we might try an even higher number of repetitions.

Reply:
**Race is:**
```
WARNING: ThreadSanitizer: data race (pid=3970251)
  Write of size 8 at 0x7b5400000cb0 by main thread:
    #0 operator delete(void*) /home/engshare/third-party2/gcc/9.x/src/gcc-10.x/libsanitizer/tsan/tsan_new_delete.cc:125 (libtsan.so.0+0x8be82)
    #1 rocksdb::SpecialEnv::~SpecialEnv() db/db_test_util.h:114 (db_sst_test+0x529385)
    #2 rocksdb::DBTestBase::~DBTestBase() db/db_test_util.cc:119 (db_sst_test+0x51547c)
    #3 rocksdb::DBSSTTest::~DBSSTTest() db/db_sst_test.cc:19 (db_sst_test+0x4e1ac5)
    #4 rocksdb::DBSSTTest_DBWithMaxSpaceAllowedWithBlobFiles_Test::~DBSSTTest_DBWithMaxSpaceAllowedWithBlobFiles_Test() db/db_sst_test.cc:1069 (db_sst_test+0x4e1ac5)
    #5 rocksdb::DBSSTTest_DBWithMaxSpaceAllowedWithBlobFiles_Test::~DBSSTTest_DBWithMaxSpaceAllowedWithBlobFiles_Test() db/db_sst_test.cc:1069 (db_sst_test+0x4e1ac5)
.....

Previous read of size 8 at 0x7b5400000cb0 by thread T4 (mutexes: write M1346):
    #0 rocksdb::EnvWrapper::GetFreeSpace(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long*) include/rocksdb/env.h:1581 (db_sst_test+0x4dff0e)
    #1 GetFreeSpace env/env.cc:557 (db_sst_test+0x89f118)
    #2 rocksdb::SstFileManagerImpl::ClearError() file/sst_file_manager_impl.cc:264 (db_sst_test+0x926187)
.....

 Mutex M1346 (0x7b5c00000738) created at:
    #0 pthread_mutex_init /home/engshare/third-party2/gcc/9.x/src/gcc-10.x/libsanitizer/tsan/tsan_interceptors.cc:1183 (libtsan.so.0+0x49677)
    #1 rocksdb::port::Mutex::Mutex(bool) port/port_posix.cc:75 (db_sst_test+0x9dbb8f)
    #2 rocksdb::SstFileManagerImpl::SstFileManagerImpl(std::shared_ptr<rocksdb::SystemClock> const&, std::shared_ptr<rocksdb::FileSystem> const&, std::shared_ptr<rocksdb::Logger> const&, long, double, unsigned long) file/sst_file_manager_impl.cc:42 (db_sst_test+0x926d89)
    #3 rocksdb::NewSstFileManager(rocksdb::Env*, std::shared_ptr<rocksdb::FileSystem>, std::shared_ptr<rocksdb::Logger>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, long, bool, rocksdb::Status*, double, unsigned long) file/sst_file_manager_impl.cc:477 (db_sst_test+0x927ca6)
    #4 rocksdb::NewSstFileManager(rocksdb::Env*, std::shared_ptr<rocksdb::Logger>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, long, bool, rocksdb::Status*, double, unsigned long) file/sst_file_manager_impl.cc:462 (db_sst_test+0x928569)
.....

 Thread T4 (tid=3970364, finished) created by thread T2 at:
    #0 pthread_create /home/engshare/third-party2/gcc/9.x/src/gcc-10.x/libsanitizer/tsan/tsan_interceptors.cc:964 (libtsan.so.0+0x613c3)
    #1 __gthread_create /home/engshare/third-party2/libgcc/9.x/src/gcc-9.x/x86_64-facebook-linux/libstdc++-v3/include/x86_64-facebook-linux/bits/gthr-default.h:663 (libstdc++.so.6+0xd993c)
    #2 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) ../../../.././libstdc++-v3/src/c++11/thread.cc:135 (libstdc++.so.6+0xd993c)
    #3 rocksdb::ErrorHandler::RecoverFromNoSpace() db/error_handler.cc:531 (db_sst_test+0x71df35)
    #4 rocksdb::ErrorHandler::SetBGError(rocksdb::Status const&, rocksdb::BackgroundErrorReason) db/error_handler.cc:355 (db_sst_test+0x71e89f)
    #5 rocksdb::BlobFileCompletionCallback::OnBlobFileCompleted(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int, unsigned long, rocksdb::BlobFileCreationReason, rocksdb::Status const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, unsigned long, unsigned long) db/blob...

```

So I think in the test we try to delete the env but other thread is accessing it which is causing the race. I will try the gtest_repeat option to run it and see if its reproducible locally first in order to test the fix.